From: "Andrey Borzenkov" <arvidjaar@mail.ru>
Subject: PATCH: 2.4.20-2mdk: ide-proc && ide_scan cleanup
To: quintela@mandrakesoft.com
Cc: cooker@linux-mandrake.com
Date: Sat, 04 Jan 2003 16:02:03 +0300
X-Mailer: mPOP Web-Mail 2.19
Reply-To: "Andrey Borzenkov" <arvidjaar@mail.ru>

Again from very old backlog.

This fixes my old error. The patch is against 2.4.20-2mdk, but it reverts DI92, so Juan, if you intend to ever apply it :) just remove DI92 and skip drivers/ide/ide.c chunk.

The patch fixes driver changing by writing into /proc/ide/hdX/driver. It does work to some extent, but current code does not null-terminate written driver name so ide_scan_device does not (always) find it.

DI92 must be removed in any case (it is buggy); this patch is optional as probably nobody ever tried to play with IDE drivers this way :)) I repost it so it is not lost. 

-andrey
[2. text/plain; 2.4.20-2mdk.ide-proc-write-driver.patch]

diff -uNp t8/drivers/ide/ide-proc.c.orig t8/drivers/ide/ide-proc.c
--- t8/drivers/ide/ide-proc.c.orig	2003-01-16 16:59:40.000000000 +0100
+++ t8/drivers/ide/ide-proc.c	2003-01-16 17:00:47.000000000 +0100
@@ -657,14 +657,23 @@ int proc_ide_read_driver
 
 EXPORT_SYMBOL(proc_ide_read_driver);
 
+#define DRIVER_REQ_SIZE (sizeof(((ide_drive_t *)0)->driver_req))
 int proc_ide_write_driver
 	(struct file *file, const char *buffer, unsigned long count, void *data)
 {
 	ide_drive_t	*drive = (ide_drive_t *) data;
+	size_t		s = count < DRIVER_REQ_SIZE ? count : DRIVER_REQ_SIZE-1;
+	char		tmpbuf[DRIVER_REQ_SIZE];
 
 	if (!capable(CAP_SYS_ADMIN))
 		return -EACCES;
-	if (ide_replace_subdriver(drive, buffer))
+	/*
+	 * Driver name may be less than 9 characters
+	 * Make sure it is properly NULL-terminated
+	 */
+	strncpy(tmpbuf, buffer, s);
+	tmpbuf[s] = '\0';
+	if (ide_replace_subdriver(drive, tmpbuf))
 		return -EINVAL;
 	return count;
 }
